examples/step-62/step-62.cc

```diff
@@ -0,0 +1,1467 @@
+
+    FEValues<dim>      fe_values(fe,
+                            quadrature_formula,
+                            update_values | update_gradients |
+                              update_quadrature_points | update_JxW_values);
+    const unsigned int dofs_per_cell = fe.dofs_per_cell;
+
+    FullMatrix<std::complex<double>> cell_matrix(dofs_per_cell, dofs_per_cell);
+    Vector<std::complex<double>>     cell_rhs(dofs_per_cell);
+
+    std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+    // Here we store the value of the right hand side, rho and the PML.
+    std::vector<Vector<double>> rhs_values(n_q_points, Vector<double>(dim));
+    std::vector<double>         rho_values(n_q_points);
+    std::vector<Vector<std::complex<double>>> pml_values(
+      n_q_points, Vector<std::complex<double>>(dim));
+
+    // We calculate the stiffness tensor for the $\lambda$ and $\mu$ that has
+    // been defined in the jupyter notebook. Note that contrary to $\rho$ the
+    // stiffness is constant among for the whole domain.
+    const SymmetricTensor<4, dim> stiffness_tensor =
+      get_stiffness_tensor<dim>(parameters.lambda, parameters.mu);
+
+    // We use the same method of step-20 for vector-valued problems.
+    const FEValuesExtractors::Vector displacement(0);
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        if (cell->is_locally_owned())
+          {
+            cell_matrix = 0;
+            cell_rhs    = 0;
+
+            // We have to calculate the values of the right hand side, rho and
+            // the PML only if we are going to calculate the mass and the
+            // stiffness matrices. Otherwise we can skip this calculation which
+            // considerably reduces the total calculation time.
+            if (calculate_quadrature_data)
+              {
+                fe_values.reinit(cell);
+
+                parameters.right_hand_side.vector_value_list(
+                  fe_values.get_quadrature_points(), rhs_values);
+                parameters.rho.value_list(fe_values.get_quadrature_points(),
+                                          rho_values);
+                parameters.pml.vector_value_list(
+                  fe_values.get_quadrature_points(), pml_values);
+              }
+
+            // We have done this in step-18. Get a pointer to the quadrature
+            // cache data local to the present cell, and, as a defensive
+            // measure, make sure that this pointer is within the bounds of the
+            // global array:
+            QuadratureCache<dim> *local_quadrature_points_data =
+              reinterpret_cast<QuadratureCache<dim> *>(cell->user_pointer());
+            Assert(local_quadrature_points_data >= &quadrature_cache.front(),
+                   ExcInternalError());
+            Assert(local_quadrature_points_data < &quadrature_cache.back(),
+                   ExcInternalError());
```

Comment:
Shouldn't the comparison be `<=`? I'm uncertain how this works...

Reply:
I'm not sure either how it works. I copied the code from step-18. I replaced `>=` by `<=` and the program complains in debug mode, thus `>=` seems to be correct.

Reply:
The `>= &quadrature_cache.front()` seems correct to me, but since `back()` is an actual element (and not the past-the-end element), I would have expected that the correct second condition is `<= &quadrature_cache.back()` rather than `< &quadrature_cache.back()`. What happens if you make this change?

Reply:
Ah, I understand now: We resize this array to
```
  quadrature_cache.resize(our_cells * quadrature_formula.size(),
                                           QuadratureCache<dim>(fe.dofs_per_cell));
```
so for every cell there are `quadrature_formula.size()` elements. We get a pointer to the first of these elements above, and so as long as `quadrature_formula.size() > 1`, the pointer we obtain is never a pointer to the last element of the array. So the assertion succeeds, but it should fail if one had a 1-point quadrature_formula. Indeed, I've tried this out and that's what happens. I'll fix step-18. Can you make the same fix here?

Reply:
See #7960.